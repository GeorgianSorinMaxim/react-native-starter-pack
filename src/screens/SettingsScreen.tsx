import React, { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Linking, View, StyleSheet } from "react-native";
import DeviceInfo from "react-native-device-info";
import uuid from "react-native-uuid";

import {
  BodyText,
  ButtonWithIcon,
  Divider,
  Title,
  Screen,
} from "../components";
import Colors from "../constants/Colors";

import { authActions } from "../store/actions/auth";
import { getUser } from "../store/selectors";

import { contact } from "../utils/contact";

import { StringValues } from "../constants/StringValues";

export const SettingsScreen = () => {
  const dispatch = useDispatch();

  const user = useSelector(getUser);

  const [bundleId, setBundleId] = useState<string>("");
  const [buildNumber, setBuildNumber] = useState<string>("");
  const [version, setVersion] = useState<string>("");
  const [readableVersion, setReadableVersion] = useState<string>("");
  const [systemName, setSystemName] = useState<string>("");
  const [ip, setIp] = useState<string>("");
  const [osVersion, setOsVersion] = useState<string>("");
  const [uuidValue, setUuid] = useState<string>("");

  const getDeviceData = async () => {
    try {
      const bundleId = await DeviceInfo.getBundleId();
      const buildNumber = await DeviceInfo.getBuildNumber();
      const version = await DeviceInfo.getVersion();
      const readableVersion = await DeviceInfo.getReadableVersion();
      const systemName = await DeviceInfo.getSystemName();
      const ip = await DeviceInfo.getIpAddress();
      const osVersion = await DeviceInfo.getSystemVersion();
      const uuidVal = uuid.v1();

      setBundleId(bundleId);
      setBuildNumber(buildNumber);
      setVersion(version);
      setReadableVersion(readableVersion);
      setSystemName(systemName);
      setIp(ip);
      setOsVersion(osVersion);
      setUuid(uuidVal.toString());
    } catch (err) {
      return console.log("getDeviceData error:", err);
    }
  };

  useEffect(() => {
    getDeviceData();
  }, []);

  const onLogout = () => {
    dispatch(authActions.logoutStart());
  };

  const onContact = () => {
    const EMAIL = "hey@devxldn.com";
    const SUBJECT = "I have a question or concern:";
    const BODY = `Please write your question or concern below:\n
      \n\n\n
      ----------------------------\n
      App ID: #${bundleId} \n
      App build number: ${buildNumber} \n
      App version: ${version} \n
      System: ${systemName} \n
      OS: ${osVersion} \n
      IP: ${ip} \n
      UUID: ${uuidValue} \n
    `;
    contact(EMAIL, SUBJECT, BODY);
  };

  const onTermsPress = () => {
    try {
      Linking.openURL("http://www.devxldn.com");
    } catch (err) {
      console.log("Linking error: ", err);
    }
  };

  return (
    <Screen noHorizontalPadding={true}>
      <View style={styles.container}>
        <Title label={StringValues.yourDetails} />
        <Divider />
        <View style={styles.bodyContainer}>
          {user?.lastName ? (
            <View style={styles.textContainer}>
              <BodyText style={styles.headerText}>Name</BodyText>
              <BodyText style={styles.contentText}>
                {user?.firstName} {user?.lastName}
              </BodyText>
            </View>
          ) : null}
          <View style={styles.textContainer}>
            <BodyText style={styles.headerText}>{StringValues.email}</BodyText>
            <BodyText style={styles.contentText}>{user?.email}</BodyText>
          </View>
          <View style={styles.textContainer}>
            <BodyText style={styles.headerText}>
              {StringValues.appVersion}
            </BodyText>
            <BodyText style={styles.contentText}>{readableVersion}</BodyText>
          </View>
          <View style={styles.textContainer}>
            <BodyText style={styles.headerText}>{StringValues.uuid}</BodyText>
            <BodyText style={styles.contentText}>{uuidValue}</BodyText>
          </View>
        </View>
      </View>

      <ButtonWithIcon
        icon="link"
        label={StringValues.termsAndConditions}
        onPress={onTermsPress}
      />
      <ButtonWithIcon
        icon="mail-outline"
        label={StringValues.contact}
        onPress={onContact}
      />
      <ButtonWithIcon
        icon="log-out-outline"
        label={StringValues.logout}
        onPress={onLogout}
      />
    </Screen>
  );
};

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: 24,
  },
  bodyContainer: {
    padding: 24,
  },
  textContainer: {
    paddingVertical: 10,
  },
  headerText: {
    paddingBottom: 4,
    color: Colors.grey,
  },
  contentText: {
    color: Colors.gold,
    fontWeight: "bold",
  },
});
