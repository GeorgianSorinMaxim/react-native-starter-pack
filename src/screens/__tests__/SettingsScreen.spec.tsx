import * as React from "react";
import renderer from "react-test-renderer";

import { SettingsScreenBase } from "../SettingsScreen";

jest.mock("@react-navigation/stack");
jest.mock("@react-navigation/native");
jest.mock("@react-navigation/bottom-tabs");

describe("SettingsScreen", () => {
  it(`renders the SettingsScreen screen`, () => {
    const defaultProps = {
      navigation: {
        setParams: jest.fn(),
        navigate: jest.fn(),
      },
      logout: jest.fn(),
      user: {
        id: "123",
        email: "test@test.com",
        firstName: "John",
        lastName: "Doe",
      },
    };

    // @ts-ignore
    const tree = renderer.create(<SettingsScreenBase {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
