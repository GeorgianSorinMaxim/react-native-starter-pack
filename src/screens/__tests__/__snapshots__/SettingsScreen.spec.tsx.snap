// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsScreen renders the SettingsScreen screen 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFF",
        "flex": 1,
        "paddingBottom": 24,
      },
    ]
  }
>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardShouldPersistTaps="always"
    keyboardSpace={0}
    onScroll={[Function]}
    resetKeyboardSpace={[Function]}
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    style={
      Object {
        "flex": 1,
        "width": "100%",
      }
    }
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <View>
      <View
        style={
          Object {
            "paddingHorizontal": 24,
          }
        }
      >
        <View
          style={
            Object {
              "marginVertical": 24,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#D4AF37",
                "fontSize": 24,
                "fontWeight": "bold",
              }
            }
          >
            Your details
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "borderBottomWidth": 0,
                "borderColor": "#D4AF37",
                "borderLeftWidth": 0,
                "borderRightWidth": 0,
                "borderTopWidth": 0.5,
              },
              null,
            ]
          }
        />
        <View
          style={
            Object {
              "padding": 24,
            }
          }
        >
          <View
            style={
              Object {
                "paddingVertical": 10,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#BABABA",
                    "paddingBottom": 4,
                  },
                ]
              }
            >
              Name
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#D4AF37",
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              John
               
              Doe
            </Text>
          </View>
          <View
            style={
              Object {
                "paddingVertical": 10,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#BABABA",
                    "paddingBottom": 4,
                  },
                ]
              }
            >
              Email
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#D4AF37",
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              test@test.com
            </Text>
          </View>
          <View
            style={
              Object {
                "paddingVertical": 10,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#BABABA",
                    "paddingBottom": 4,
                  },
                ]
              }
            >
              App version
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#D4AF37",
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              
            </Text>
          </View>
          <View
            style={
              Object {
                "paddingVertical": 10,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#BABABA",
                    "paddingBottom": 4,
                  },
                ]
              }
            >
              UUID
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#D4AF37",
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
      <RNGestureHandlerButton
        activeOpacity={0.105}
        collapsable={false}
        onActiveStateChange={[Function]}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        onPress={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderColor": "#BABABA",
              "borderWidth": 0.5,
              "marginBottom": 1,
              "marginTop": -1,
              "paddingHorizontal": 15,
              "paddingVertical": 15,
            },
          ]
        }
        underlayColor="black"
      >
        <View
          style={
            Object {
              "backgroundColor": "black",
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "marginRight": 12,
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#D4AF37",
                    "fontSize": 18,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Ionicons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
          <Text
            style={
              Object {
                "alignSelf": "flex-start",
                "fontSize": 12,
                "letterSpacing": 4,
                "marginTop": 1,
                "textTransform": "uppercase",
              }
            }
          >
            Terms & Conditions
          </Text>
        </View>
      </RNGestureHandlerButton>
      <RNGestureHandlerButton
        activeOpacity={0.105}
        collapsable={false}
        onActiveStateChange={[Function]}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        onPress={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderColor": "#BABABA",
              "borderWidth": 0.5,
              "marginBottom": 1,
              "marginTop": -1,
              "paddingHorizontal": 15,
              "paddingVertical": 15,
            },
          ]
        }
        underlayColor="black"
      >
        <View
          style={
            Object {
              "backgroundColor": "black",
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "marginRight": 12,
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#D4AF37",
                    "fontSize": 18,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Ionicons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
          <Text
            style={
              Object {
                "alignSelf": "flex-start",
                "fontSize": 12,
                "letterSpacing": 4,
                "marginTop": 1,
                "textTransform": "uppercase",
              }
            }
          >
            Contact
          </Text>
        </View>
      </RNGestureHandlerButton>
      <RNGestureHandlerButton
        activeOpacity={0.105}
        collapsable={false}
        onActiveStateChange={[Function]}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        onPress={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderColor": "#BABABA",
              "borderWidth": 0.5,
              "marginBottom": 1,
              "marginTop": -1,
              "paddingHorizontal": 15,
              "paddingVertical": 15,
            },
          ]
        }
        underlayColor="black"
      >
        <View
          style={
            Object {
              "backgroundColor": "black",
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "marginRight": 12,
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#D4AF37",
                    "fontSize": 18,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Ionicons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
          <Text
            style={
              Object {
                "alignSelf": "flex-start",
                "fontSize": 12,
                "letterSpacing": 4,
                "marginTop": 1,
                "textTransform": "uppercase",
              }
            }
          >
            Logout
          </Text>
        </View>
      </RNGestureHandlerButton>
    </View>
  </RCTScrollView>
</View>
`;
